flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/ExtractionUtils.java

```diff
@@ -348,6 +350,53 @@ public static Optional<Class<?>> extractSimpleGeneric(
         }
     }
 
+    /** Resolves a variable type while accepting a context for resolution. */
+    public static Type resolveVariableWithClassContext(@Nullable Type contextType, Type type) {
+        final List<Type> typeHierarchy;
+        if (contextType != null) {
+            typeHierarchy = collectTypeHierarchy(contextType);
+        } else {
+            typeHierarchy = Collections.emptyList();
+        }
+        if (!containsTypeVariable(type)) {
```

Comment:
not sure if this was intended but with this check you exclude the `WildcardType`:
```
eval(CompletableFuture<? extends Integer> s)
```
which is correct because the if clause further down could not handle it, I would at least add a comment to the false in `containsTypeVariable`

Reply:
I did exclude that intentionally.  I don't think it would be easy to call back on the future with a concrete type even if such a `CompletableFuture<? extends Number>` were a parameter since I think this is a compile time error. I tried some examples of weird method overriding to allow calling back on that future and got one that worked, but it's awkward (and throws a reasonable error when you run it anyhow). Will add a comment in containsTypeVariable about keeping in sync with this method.